### PR TITLE
Add support for parsing unicode in the units module

### PR DESCRIPTION
**Pull request by MaxNoe**
### Description

As a first simple step in adding feature #9347, this PR adds two unicode symbols
as possible representations for the micro prefix `MICRO SIGN` and `GREEK SMALL LETTER MU`.

While the second one is strictly speaking not correct, it probably appears often enough in the wild to be useful to support.

### CODE REVIEW

**Comment by MaxNoe** [comment]
I though about which representation would be best, basically there are three options:
Just inserting the unicode character: `'µ'`, using the code point `'\u00b5'` or the name of the character `'\N{MICRO SIGN}'`.

I went for the most verbose one, as it is both ascii in the code and human readable. Any other opinions on this?

**Comment by pllim** [comment]
If it is "strictly speaking not correct," should we even support it? :thinking:

**Comment by MaxNoe** [comment]
Semantically correct would always be to use 'MICRO SIGN', however I think as 
* It's very common to use 'GREEK SMALL LETTER MU'
* It's not a confflict with anything
* The two characters look the same in very many fonts
it's probably better to add both

**Comment by mhvk** [comment]
The PR looks OK, but I'm a bit with @pllim on whether we should support the greek letter as well. My sense would be that it is OK to interpret it on input, but that we should not expose it in `astropy.units`. I think the easiest way to do this is to, inside the parser, convert one to another.  I'll try to see if I can make a quick PR to your PR.

**Comment by MaxNoe** [comment]
I see the potential of "Why does µ work but μ not"? I don't really see a harm of supporting both as long as `u.µm == u.μm`

**Comment by pllim** [comment]
They do look very similar... :eyes:

**Comment by mhvk** [comment]
My main hesitation is creating yet more entries in the units registry - indeed, I'm not completely sold on even having it there (but very sold on being able to parse it; I think we should roundtrip in unicode as well).

How about we start simple and just allow the mu in strings? I made a PR to your PR...

**Comment by mhvk** [comment]
p.s.  Off for a bit, back in about 1.5 hours.

**Comment by MaxNoe** [comment]
I also fixed the used character for the micro sign in the unicode output format. See, @pllim commonly used :rofl:

File: astropy/units/format/generic.py [hunk 1]
```diff
@@ -148,7 +148,9 @@ def t_FUNCNAME(t):
             return t
 
         def t_UNIT(t):
-            r"%|([YZEPTGMkhdcmunpfazy]?'((?!\d)\w)+')|((?!\d)\w)+"
+            "%|([YZEPTGMkhdcmu\N{MICRO SIGN}npfazy]?"+r"'((?!\d)\w)+')|((?!\d)\w)+"
```
Comment (pllim) [inline astropy/units/format/generic.py]: Should remain a raw string, right?
```suggestion
            r"%|([YZEPTGMkhdcmu\N{MICRO SIGN}npfazy]?"+r"'((?!\d)\w)+')|((?!\d)\w)+"
```

Comment (mhvk) [reply astropy/units/format/generic.py]: No, in the first part I need `\N` to be interpreted correctly, but for the later parts one wants to keep `\d` etc. This is why I'm adding a regular to a raw string...

File: astropy/units/format/generic.py [hunk 2]
```diff
@@ -464,11 +466,37 @@ def _parse_unit(cls, s, detailed_exception=True):
         else:
             raise ValueError()
 
+    _translations = str.maketrans({
+        '\N{GREEK SMALL LETTER MU}': '\N{MICRO SIGN}',
+        '\N{MINUS SIGN}': '-',
+        '\N{SUPERSCRIPT MINUS}': '-',
+        '\N{SUPERSCRIPT PLUS SIGN}': '+',
+        '\N{SUPERSCRIPT ZERO}': '0',
```
Comment (mhvk) [inline astropy/units/format/generic.py]: I like this but it also worries me a bit - as I think one should check that there is not a mix of superscript and non-superscript numbers. So, I *think* I'd rather do this in `p_numeric_power` - allow new tokens `superscript_sign superscript_uint`  - or pre-translate but check that there are no regular numbers next to it.

File: astropy/units/format/generic.py (hunk 1 above)
Comment (pllim) [reply astropy/units/format/generic.py]: Okay, sorry for the noise!

**Comment by mhvk** [comment]
@MaxNoe - I made another PR to your branch, where superscripts are still translated, but as part of a regex that surrounds the translations with parentheses - which are recognized as powers and avoids the mix of regular and superscript numbers just working.

p.s.  I had hoped to do it with the lexer/parser itself, but did not manage.  Might still be better...

**Comment by mhvk** [comment]
p.s. It might be fun to see if the roundtrip tests that are done now also work in unicode!

p.s.2 If there are superscript slashes and periods, possibly we should add those too...

File: astropy/units/format/generic.py (hunk 2 above)
Comment (MaxNoe) [reply astropy/units/format/generic.py]: merged your PR

**Comment by MaxNoe** [comment]
Ok, Forgot about Angstom, Ohm and degrees

**Comment by MaxNoe** [comment]
Ok, that should be all, please review!

**Comment by MaxNoe** [comment]
Tests should be passing now. I did not try to test these roundtrips. How are those support to work?

File: astropy/units/tests/test_format.py [hunk 3]
```diff
@@ -569,3 +570,35 @@ def test_powers(power, expected):
...
+    ('m\N{SUPERSCRIPT ONE}\N{SUPERSCRIPT ZERO}', u.m**10),
+    ('\N{LATIN CAPITAL LETTER A WITH RING ABOVE}', u.angstrom),
+    ('\N{GREEK CAPITAL LETTER OMEGA}', u.ohm),
+    ('\N{OHM SIGN}', u.ohm), # deprecated but for compatibility
+    ('°C', u.deg_C),
+    ('°', u.deg),
+])
+def test_unicode(string, unit):
+    assert u_format.Generic.parse(string) == unit
+    assert u.Unit(string) == unit
```
Comment (aarchiba) [inline astropy/units/tests/test_format.py]: I don't see it in the tests - will I be able to write u.Å in my source code?

Comment (MaxNoe) [reply astropy/units/tests/test_format.py]: Ah yes. At the moment `u.Å` and `u.Ω` work. Micro does not.

@mhvk should we add the micro sign?

**Comment by mhvk** [comment]
@MaxNoe - my sense would be *not* to change anything in the unit registry in this PR, and thus also not add to `u.<whatever>` - I think adding non-ascii items to any module is something that needs a bit larger discussion (if only because of the annoyance of the different mu; if we can do Ohm, I definitely want micro-Ohm too!). So, I would suggest to move the Angstrom and Ohm changes from this PR to a new one that we can discuss more broadly.

p.s. I do like the degree additions. Of course, this begs the question about arcsecond and arcminute - this is where the round-trip tests would come in. One would hope one could work from the tests that now use `RoundtripBase`.

File: astropy/units/tests/test_format.py (hunk 3 above)
Comment (MaxNoe) [reply astropy/units/tests/test_format.py]: removed again according to @mhvk

**Comment by MaxNoe** [comment]
Ok, so I added support for just parsin Ohm back in because it was easy, support for Å is harder, so I left it out.

Two reasons:
* You cannot simply replace Å with A since that would be ampere in case it's only `Å` and not `Ångström`
* There are two possible unicode reprensations of Å, the character itself and A plus the combining °.
This would create the necessity of unicode normalization or supporting both.

File: astropy/units/format/generic.py [hunk 4]
```diff
@@ -469,8 +476,8 @@ def _parse_unit(cls, s, detailed_exception=True):
     _translations = {
         '\N{GREEK SMALL LETTER MU}': '\N{MICRO SIGN}',
         '\N{MINUS SIGN}': '-',
-        '\N{OHM SIGN}': 'Ohm',
-        '\N{GREEK CAPITAL LETTER OMEGA}': 'Ohm',
+        '\N{GREEK CAPITAL LETTER OMEGA}': '\N{OHM SIGN}',
+        '\N{LATIN CAPITAL LETTER A WITH RING ABOVE}': '\N{ANGSTROM SIGN}'
```
Comment (MaxNoe) [inline astropy/units/format/generic.py]: Not the other way around?

Comment (MaxNoe) [reply astropy/units/format/generic.py]: I read that both these specialized versions are considered deprecated (at least the OHM SIGN) and should not be used anymore.

Comment (MaxNoe) [reply astropy/units/format/generic.py]: See https://en.wiktionary.org/wiki/%CE%A9

Comment (mhvk) [reply astropy/units/format/generic.py]: I thought inside `_parse_unit`, it was clearer if we were comparing with the more obvious 'OHM SIGN' and 'ANGSTROM SIGN'.

**Comment by MaxNoe** [comment]
@mhvk sure fine!

There is also `unicodedata.normalize` in the standard lib which takes care of composed versus single characters.

File: astropy/units/format/generic.py [hunk 5]
```diff
@@ -464,15 +473,72 @@ def _parse_unit(cls, s, detailed_exception=True):
...
+            # Translate possible degrees.
+            s = cls._regex_deg.sub(cls._convert_deg, s)
 
         result = cls._do_parse(s, debug=debug)
         # Check for excess solidi, but exclude fractional exponents (accepted)
-        if (s.count('/') > 1 and
-                s.count('/') - len(re.findall(r'\(\d+/\d+\)', s)) > 1):
+        n_slashes = s.count('/')
+        n_fractions = len(re.findall(r'\(\d+/\d+\)', s))
+        if n_slashes > 1 and (n_slashes - n_fractions) > 1:
```
Comment (mhvk) [inline astropy/units/format/generic.py]: Why this change? You now always look for fractions, which will slow down the parsing for all units. I'd at least keep it inside an `if` statement.

File: astropy/units/format/generic.py [hunk 6]
```diff
@@ -454,7 +455,22 @@ def _parse_unit(cls, s, detailed_exception=True):
...
         if s == '%':
             return registry['percent']
-        elif s in registry:
+
+        if not s.isascii():
+            # common normalization of unicode strings to avoid
+            # having to deal with multiple representations of
+            # the same character. This normalizes to "composed" form
+            # and will e.g. convert OHM SIGN to GREEK CAPITAL LETTER OMEGA
+            s = unicodedata.normalize('NFC', s)
```
Comment (mhvk) [inline astropy/units/format/generic.py]: Do we still need that here, given that it is done in `parse` already? I'd move the comment there too.

**Comment by MaxNoe** [comment]
@mhvk unfortunately `isascii` is new in python 3.7.

File: astropy/units/format/generic.py [hunk 7]
```diff
@@ -464,15 +486,74 @@ def _parse_unit(cls, s, detailed_exception=True):
         else:
             raise ValueError()
 
+    _translations = {
+        '\N{GREEK SMALL LETTER MU}': '\N{MICRO SIGN}',
+        '\N{MINUS SIGN}': '-',
+    }
+    # key in the translations dict must be the codepoint
+    _translations = {ord(k):v for k, v in _translations.items()}
```
Comment (mhvk) [inline astropy/units/format/generic.py]: Doesn't really matter, but why not use `str.maketrans` any more?

**Comment by mhvk** [comment]
Beyond the one last comment this all looks great!

**Comment by mhvk** [comment]
I just made the small change. I think it is all OK now, but since I'm arguably now a co-committer, probably best to get an independent sign-off. @adrn or @dhomeier - might you be able to review?

**Comment by adrn** [comment]
This looks nice! Unfortunately I won't be able to give it a thorough review until ~Tuesday next week, as I'm away with family for the long (US) weekend.

**Comment by dhomeier** [comment]
I haven't had time for a full review, but tested it under Python 3.8; from a quick overview looks good to me too - @mhvk's last commit certainly does.
One minor question - is it policy not to support prefixed Ångström variants? It's understandable, since the base unit is already deprecated, but with the unicode versions supported, one might more likely encounter `u.Unit('mÅ')` as well.

**Comment by mhvk** [comment]
For `mÅ`, I'm not in favour either, but also not really against; my sense would be to let it be driven by user request (i.e., @dhomeier, if you would like to see it, it's fine by me).

**Comment by mhvk** [comment]
@adrn - would you still be able to have a look? With @dhomeier's quicker review and @taldcroft's liking of the general idea in https://github.com/astropy/astropy/issues/9347#issuecomment-541242918, I think there has been enough review to merge, but a good look is always helpful.

**Comment by dhomeier** [comment]
> For `mÅ`, I'm not in favour either, but also not really against; my sense would be to let it be driven by user request (i.e., @dhomeier, if you would like to see it, it's fine by me).

I am not strongly pushing for it either, but in the spirit of the earlier discussion on which µ variants to accept, I think it's worth at looking at the potential harm vs. benefits. Stuff copied from older literature might very well come in milliangstroms, too.

**Comment by mhvk** [comment]
True. Though note that this is somewhat orthogonal to the discussion here - one would just set `prefixes=['m']` in the definition of `Angstrom` and things would just work (just like `mOhm` would now).

**Comment by bsipocz** [comment]
@mhvk or @adrn - would either of you approve and merge this now?

**Comment by mhvk** [comment]
As a co-author, I perhaps shouldn't, but if @adrn doesn't get to it, I'l, do it tomorrow.

**Comment by MaxNoe** [comment]
@mhvk Was a pleisure, thanks for all the support!
